src/audio/smart_amp/smart_amp.c

```diff
@@ -0,0 +1,668 @@
+
+	/* buffer : feed forward process output */
+	size = DSM_FF_BUF_SZ * sizeof(int16_t);
+	hspk->buf.output = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	if (!hspk->buf.output)
+		goto err;
+	mem_sz += size;
+
+	/* buffer : feedback voltage */
+	size = DSM_FF_BUF_SZ * sizeof(int16_t);
+	hspk->buf.voltage = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	if (!hspk->buf.voltage)
+		goto err;
+	mem_sz += size;
+
+	/* buffer : feedback current */
+	size = DSM_FF_BUF_SZ * sizeof(int16_t);
+	hspk->buf.current = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	if (!hspk->buf.current)
+		goto err;
+	mem_sz += size;
+
+	/* buffer : feed forward variable length -> fixed length */
+	size = DSM_FF_BUF_DB_SZ * sizeof(int32_t);
+	hspk->buf.ff.buf = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	if (!hspk->buf.ff.buf)
+		goto err;
+	mem_sz += size;
+
+	/* buffer : feed forward variable length <- fixed length */
+	size = DSM_FF_BUF_DB_SZ * sizeof(int32_t);
+	hspk->buf.ff_out.buf = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	if (!hspk->buf.ff_out.buf)
+		goto err;
+	mem_sz += size;
+
+	/* buffer : feedback variable length -> fixed length */
+	size = DSM_FB_BUF_DB_SZ * sizeof(int32_t);
+	hspk->buf.fb.buf = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	if (!hspk->buf.fb.buf)
+		goto err;
+	mem_sz += size;
+
+	comp_dbg(dev, "[DSM] module:%p (%d bytes used)",
+		 (uintptr_t)hspk, mem_sz);
+
+	return 0;
+err:
+	smart_amp_free_memory(sad, dev);
+	return -ENOMEM;
+}
+
+static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
+				      struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *sa;
+	struct sof_ipc_comp_process *ipc_sa =
+		(struct sof_ipc_comp_process *)comp;
+	struct smart_amp_data *sad;
+	struct sof_smart_amp_config *cfg;
+	size_t bs;
+	int ret;
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	sad = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*sad));
+
+	if (!sad) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, sad);
+
+	sa = COMP_GET_IPC(dev, sof_ipc_comp_process);
+
+	ret = memcpy_s(sa, sizeof(*sa), ipc_sa,
+		       sizeof(struct sof_ipc_comp_process));
+	assert(!ret);
+
+	cfg = (struct sof_smart_amp_config *)ipc_sa->data;
+	bs = ipc_sa->size;
+
+	if (bs > 0 && bs < sizeof(struct sof_smart_amp_config)) {
+		comp_err(dev, "smart_amp_new(): failed to apply config");
+
+		if (sad)
+			rfree(sad);
+		rfree(sad);
+		return NULL;
+	}
+
+	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg, bs);
+
+	if (smart_amp_alloc_memory(sad, dev) != 0)
+		return NULL;
+	if (smart_amp_init(sad->mod_handle, dev))
+		return NULL;
+
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+}
+
+static int smart_amp_set_config(struct comp_dev *dev,
+				struct sof_ipc_ctrl_data *cdata)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct sof_smart_amp_config *cfg;
+	size_t bs;
+
+	/* Copy new config, find size from header */
+	cfg = (struct sof_smart_amp_config *)cdata->data->data;
```

Comment:
Please take https://github.com/thesofproject/sof/pull/3233 as reference to fix the GCC 9 bulid warning.
```suggestion
	cfg = (struct sof_smart_amp_config *)
	       ASSUME_ALIGNED(cdata->data->data, sizeof(uint32_t));
```

Reply:
Thanks for the comment. Applied a fix.